source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,352 @@
+#include "common/buffer/buffer_impl.h"
+
+#include "absl/strings/str_cat.h"
+#include "fmt/printf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace PostgresProxy {
+
+/**
+ * Postgres messages are described in official Postgres documentation:
+ * https://www.postgresql.org/docs/12/protocol-message-formats.html
+ *
+ * Most of messages start with 1-byte message identifier followed by 4-bytes length field. Few
+ * messages are defined without starting 1-byte character and are used during well-defined initial
+ * stage of connection process.
+ *
+ * Messages are composed from various fields: 8, 16, 32-bit integers, String, Arrays, etc.
+ *
+ * Structures defined below have the same naming as types used in official Postgres documentation.
+ *
+ * Each structure has the following methods:
+ * read - to read number of bytes from received buffer. The number of bytes depends on structure
+ * type. toString - method returns displayable representation of the structure value.
+ *
+ */
+
+// Template for integer types.
+// Size of integer types is fixed and depends on the type of integer.
+template <typename T> class Int {
+public:
+  /**
+   * Read integer value from data buffer.
+   * @param data reference to a buffer containing data to read.
+   * @param pos offset in the buffer where data to read is located. Successful read will advance
+   * this parameter.
+   * @param left number of bytes to be read to reach the end of Postgres message.
+   * Successful read will adjust this parameter.
+   * @return boolean value indicating whether read was successful. If read returns
+   * false "pos" and "left" params are not updated. When read is not successful,
+   * the caller should not continue reading next values from the data buffer
+   * for the current message.
+   */
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((data.length() - pos) < sizeof(T)) {
+      return false;
+    }
+    value_ = data.peekBEInt<T>(pos);
+    pos += sizeof(T);
+    left -= sizeof(T);
+    return true;
+  }
+
+  std::string toString() const { return fmt::format(getFormat(), value_); }
+
+  constexpr const char* getFormat() const { return "[{}]"; }
+  T get() const { return value_; }
+
+private:
+  T value_{};
+};
+
+using Int32 = Int<uint32_t>;
+using Int16 = Int<uint16_t>;
+using Int8 = Int<uint8_t>;
+
+// 8-bits character value.
+using Byte1 = Int<char>;
+
+// String type requires byte with zero value to indicate end of string.
+class String {
```

Comment:
to me, all the implementations in this file (except the template one) seem like they should not be inlined into the header.

That means they need to be recompiled every time this .h is included, and will bloat debug info, etc.

Suggest having these just be declarations and putting the impls in a .cc

Reply:
Done. Moved all non-template methods to cc file.